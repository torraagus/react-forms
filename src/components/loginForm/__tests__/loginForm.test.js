import React from "react";
import ReactDOM from "react-dom";

import { render, cleanup, fireEvent, act } from "@testing-library/react";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom/extend-expect";

import LoginForm from "../loginForm.component";

beforeEach(cleanup);

it("should render without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<LoginForm />, div);
});

it("matches snapshot", () => {
  const tree = renderer.create(<LoginForm />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("calls onSubmit with the email and password when submit is clicked", async () => {
  const { getByTestId } = render(<LoginForm />);
  const user = {email: 'test@gmail.com', password: 'passtest'}
  
  await act( async () => {
    fireEvent.change(getByTestId("email"), { target: { value: user.email } });
  });
  expect(getByTestId("email").value).toBe(user.email);

  await act( async () => {
    fireEvent.change(getByTestId("password"), {
      target: { value: user.password },
    });
  });
  expect(getByTestId("password").value).toBe(user.password);
  
  await act( async () => {
    fireEvent.submit(getByTestId("form"));
  });
  expect(getByTestId("emailErrors").textContent).toBe("");
  expect(getByTestId("passwordErrors").textContent).toBe("");
});

test("shows an error message when submit is clicked and no email is provided", async () => {
  const { getByTestId } = render(<LoginForm />);

  await act( async () => {
    fireEvent.submit(getByTestId("form"));
  });
  expect(getByTestId("email").value).toBe("");
  expect(getByTestId("emailErrors").textContent).toBe("The email field is required");
});

test("shows an error message when submit is clicked and email format is wrong", async () => {
  const { getByTestId } = render(<LoginForm />);

  await act( async () => {
    fireEvent.change(getByTestId("email"), { target: { value: "test" } });
  });
  expect(getByTestId("email").value).toBe("test");
  
  await act( async () => {
    fireEvent.submit(getByTestId("form"));
  });
  expect(getByTestId("emailErrors").textContent).toBe("The email format is wrong");
});

test("shows an error message when submit is clicked and no password is provided", async () => {
  const { getByTestId } = render(<LoginForm />);

  await act( async () => {
    fireEvent.submit(getByTestId("form"));
  });
  expect(getByTestId("password").value).toBe("");
  expect(getByTestId("passwordErrors").textContent).toBe("The password field is required");
});

test("shows an error message when submit is clicked and the password provided have less than 8 characters", async () => {
  const { getByTestId } = render(<LoginForm />);
  
  await act( async () => {
    fireEvent.change(getByTestId("password"), { target: { value: "pass" } });
  });
  expect(getByTestId("password").value).toBe("pass");
  
  await act( async () => {
    fireEvent.submit(getByTestId("form"));
  });
  expect(getByTestId("passwordErrors").textContent).toBe("The password must have at least 8 characters");
});

test("shows the password", async () => {
  const { getByTestId } = render(<LoginForm />);

  await act( async () => {
    fireEvent.click(getByTestId("togglePassword"));
  });

  expect(getByTestId("password").type).toBe('text');
});

test("hide the password", async () => {
  const { getByTestId } = render(<LoginForm />);

  await act( async () => {
    fireEvent.click(getByTestId("togglePassword"));
  });

  await act( async () => {
    fireEvent.click(getByTestId("togglePassword"));
  });

  expect(getByTestId("password").type).toBe('password');
});