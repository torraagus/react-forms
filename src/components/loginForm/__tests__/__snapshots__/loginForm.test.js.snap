// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="mt-5 container-fluid"
>
  <div
    className="row center-xs"
  >
    <div
      className=""
    >
      <h2>
        Login
      </h2>
    </div>
  </div>
  <div
    className="row center-xs"
  >
    <div
      className=""
    >
      <p>
        Complete the form below please
      </p>
    </div>
  </div>
  <div
    className="row center-xs"
  >
    <div
      className=""
    >
      <form
        className=""
        data-testid="form"
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            data-testid="username"
            id="formBasicEmail"
            name="username"
            placeholder="Enter email"
            type="email"
          />
          <small
            className="text-muted form-text"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            data-testid="password"
            id="formBasicPassword"
            name="password"
            placeholder="Enter password"
            type="password"
          />
          <small
            className="text-muted form-text"
          />
        </div>
        <input
          defaultChecked={true}
          id="rememberMe"
          name="rememberMe"
          type="checkbox"
        />
        <label
          className="ml-2"
          htmlFor="rememberMe"
        >
          Remember me
        </label>
        <br />
        <br />
        <button
          className="btn btn-primary"
          disabled={false}
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>
  <br />
  <div
    className="row center-xs"
  >
    <div
      className=""
    />
  </div>
</div>
`;
